cowbird/permissions_synchronizer.py

```diff
@@ -269,43 +318,59 @@ def _find_matching_res(self, service_type: str, resource_nametype_path: str) ->
 
     @staticmethod
     def _create_res_data(target_segments: List[ConfigSegment],
-                         input_matched_groups: Dict[str, str],
+                         input_matched_groups: Collection[str],
                          ) -> List[ResourceSegment]:
         """
         Creates resource data, by replacing any tokens found in the segment names to their actual corresponding values.
         This data includes the name and type of each segments of a full resource path.
+        In the case where a regex is found in the target segment, the data will be formed using the same resource_type
+        for every match in the current segment.
 
         :param target_segments: List containing the name and type info of each segment of the target resource path.
         :param input_matched_groups:
         """
         res_data: List[ResourceSegment] = []
         for segment in target_segments:
-            matched_groups = re.match(NAMED_TOKEN_REGEX, segment["name"])
-            if matched_groups:
-                res_data.append({
-                    "resource_name": input_matched_groups[matched_groups.groups()[0]],
-                    "resource_type": segment["type"]
-                })
-            elif segment["name"] == MULTI_TOKEN:
-                multi_segments = input_matched_groups["multi_token"]
-                # Skip the segment if the multi_token matched 0 times, resulting in an empty string.
+            # Use the regex to create the res_data
+            if segment.get("regex") is not None:
+                regex = segment.get("regex")
+                matches = re.search(regex, input_matched_groups)
+                multi_segments = matches.group()
                 if multi_segments:
                     for seg in multi_segments.split("/"):
-                        if seg:  # Ignore empty splits
+                        if seg:
                             res_data.append({
                                 "resource_name": seg,
                                 "resource_type": segment["type"]
                             })
+
             else:
-                res_data.append({
-                    "resource_name": segment["name"],
-                    "resource_type": segment["type"]
-                })
+                matched_groups = re.match(NAMED_TOKEN_REGEX, segment["name"])
+                if matched_groups:
+                    res_data.append({
+                        "resource_name": input_matched_groups[matched_groups.groups()[0]],
+                        "resource_type": segment["type"]
+                    })
+                elif segment["name"] == MULTI_TOKEN:
+                    multi_segments = input_matched_groups["multi_token"]
```

Comment:
The way `input_matched_groups` seemed to be accessed here resembles a `dict`, but the above `regex` handling seems to expect a `collection[str]`. There might be some mismatching typings happening at some point.

Reply:
@fmigneault Previously the expected struc for input_matched_groups was `Dict[str, str]` ,  _find_matching_res would return `Tuple[str, Dict[str, str]]`  but now can returns  `Tuple[str, Dict[str, str] or str] , since I use matches.group() instead of matches.groupdict() when there's an exact match. I tried some union type to handle both case but I would get errors. `collection[str]` allowed me to return `Dict[str, str] or str`. I though of returning in both case a Dict and have a dummy key "regex_match" that is used in the other function as input_matched_groups["regex_match"]. But there will still be an issue if a regex is used in the target res_key but not in the source res_key. 

Do you have a suggestion on how to handle this edge case? Would we need to force in the config schema validator that if `regex` is found on a segment of a source res, the target res key has to have it also ?